pkg/serializer/jsonstream/compressor.go

```diff
@@ -0,0 +1,258 @@
+	if !c.firstItem {
+		uncompressedDataSize += len(jsonSeparator)
+	}
+	return compression.CompressBound(uncompressedDataSize) <= c.remainingSpace() && c.uncompressedWritten+uncompressedDataSize <= maxUncompressedSize
+}
+
+// pack flushes the temporary uncompressed buffer input to the compression writer
+func (c *compressor) pack() error {
+	expvarsTotalCycles.Add(1)
+	n, err := c.input.WriteTo(c.zipper)
+	if err != nil {
+		return err
+	}
+	c.uncompressedWritten += int(n)
+	c.zipper.Flush()
+	c.input.Reset()
+	return nil
+}
+
+// addItem will try to add the given item
+func (c *compressor) addItem(data []byte) error {
+	// check item size sanity
+	if !c.checkItemSize(data) {
+		return errTooBig
+	}
+	// check max repack cycles
+	if c.repacks >= maxRepacks {
+		return errPayloadFull
+	}
+
+	if !c.hasRoomForItem(data) {
+		if c.input.Len() == 0 {
+			return errPayloadFull
+		}
+		err := c.pack()
+		if err != nil {
+			return err
+		}
+		if !c.hasRoomForItem(data) {
+			return errPayloadFull
+		}
+		c.repacks++
+	}
+
+	// Write the separator between items
+	if c.firstItem {
+		c.firstItem = false
+	} else {
+		c.input.Write(jsonSeparator)
+	}
+
+	c.input.Write(data)
+	return nil
+}
+
+func (c *compressor) close() ([]byte, error) {
+	// Flush remaining uncompressed data
+	if c.input.Len() > 0 {
+		n, err := c.input.WriteTo(c.zipper)
+		c.uncompressedWritten += int(n)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Add json footer
+	n, err := c.zipper.Write(c.footer)
+	c.uncompressedWritten += int(n)
+	if err != nil {
+		return nil, err
+	}
+	// Add zlib footer and close
+	err = c.zipper.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	c.input.Reset()
+	inputBufferPool.Put(c.input)
+
+	expvarsTotalPayloads.Add(1)
+	expvarsBytesIn.Add(int64(c.uncompressedWritten))
+	expvarsBytesOut.Add(int64(c.compressed.Len()))
+
+	return c.compressed.Bytes(), nil
+}
+
+func (c *compressor) remainingSpace() int {
+	return maxPayloadSize - c.compressed.Len() - len(c.footer)
+}
+
+// Payloads serializes a metadata payload and sends it to the forwarder
+func Payloads(m marshaler.StreamJSONMarshaler) (forwarder.Payloads, error) {
+	var output forwarder.Payloads
+	var i int
+	itemCount := m.Len()
+	expvarsTotalCalls.Add(1)
+
+	compressor, err := newCompressor(m.JSONHeader(), m.JSONFooter())
+	if err != nil {
+		return nil, err
+	}
+
+	for i < itemCount {
+		json, err := m.JSONItem(i)
+		if err != nil {
+			log.Warnf("error marshalling an item, skipping: %s", err)
+			i++
+			continue
+		}
+
+		switch compressor.addItem(json) {
+		case errPayloadFull:
+			// payload is full, we need to create a new one
+			payload, err := compressor.close()
+			if err != nil {
+				return output, err
+			}
+			output = append(output, &payload)
+			compressor, err = newCompressor(m.JSONHeader(), m.JSONFooter())
+			if err != nil {
+				return nil, err
+			}
+		case nil:
+			// All good, continue to next item
+			i++
+			expvarsTotalItems.Add(1)
+			continue
+		default:
```

Comment:
nit: if we hit an `errTooBig`, it'd be nice to log some basic info on the item that's too big